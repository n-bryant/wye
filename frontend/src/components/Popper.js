import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";

import styles from "./Popper.styles";

export const constants = {
  POPPER_WIDTH: 300,
  POPPER_HEIGHT: 354,
  HEADER_HEIGHT: 56,
  POPPER_MARGIN: 12,
  ARROW_ORIENTATION: {
    LEFT: "left",
    RIGHT: "right"
  }
};

/**
 * determine the properties to apply to the popper
 * - based on the anchor's dimensions/positioning
 * @param {Object} anchorRect
 * @returns {Object}
 */
export const getPopperProps = anchorRect => {
  const {
    POPPER_WIDTH,
    POPPER_MARGIN,
    HEADER_HEIGHT,
    POPPER_HEIGHT,
    ARROW_ORIENTATION
  } = constants;

  // check to make sure if/where the popper has the requisite space to render
  const hasRoomRight =
    anchorRect.right + POPPER_WIDTH + POPPER_MARGIN < window.innerWidth;
  const hasRoomLeft = anchorRect.left > POPPER_WIDTH + POPPER_MARGIN;
  const hasRoomBottom = window.innerHeight - anchorRect.top > POPPER_HEIGHT;
  const hasRoomTop = window.innerHeight - HEADER_HEIGHT > POPPER_HEIGHT;

  // we don't want to render if there isn't space available
  let displayProperties = {
    style: {
      display: "none"
    }
  };
  if (!hasRoomTop) {
    return displayProperties;
  }

  // set initial arrow position
  let arrowStyle = {
    top: `calc(${anchorRect.height}px / 4)`
  };

  // set the left/right orientation of the popper relative to the anchor and based on available space
  if (hasRoomRight) {
    displayProperties = {
      arrowOrientation: ARROW_ORIENTATION.LEFT,
      arrowStyle,
      style: {
        left: `${anchorRect.width + POPPER_MARGIN}px`
      }
    };
  } else if (hasRoomLeft) {
    displayProperties = {
      arrowOrientation: ARROW_ORIENTATION.RIGHT,
      arrowStyle,
      style: {
        left: `-${POPPER_WIDTH + POPPER_MARGIN}px`
      }
    };
  }

  // calculate and apply vertical offset if there is not enough room below
  // the anchor to render the popper
  if (hasRoomBottom === false && displayProperties) {
    const offset =
      POPPER_HEIGHT -
      (window.innerHeight - anchorRect.bottom + anchorRect.height);
    displayProperties = {
      ...displayProperties,
      arrowStyle: {
        // make sure the arrow position is also adjusted
        top: `calc(${anchorRect.height}px / 4 + ${offset}px)`
      },
      style: {
        ...displayProperties.style,
        top: -offset
      }
    };
  }

  return displayProperties;
};

/**
 * renders a Popper with the provided content
 * - positioning based on screen dimensions and anchor positioning
 * - options for displaying arrow
 */
export const Popper = props => {
  const classnames = Popper.classnames(props);
  const { anchorEl, displayArrow } = props;
  const { POPPER_WIDTH, POPPER_HEIGHT, ARROW_ORIENTATION } = constants;
  const [displayProperties] = React.useState(
    getPopperProps(anchorEl.getBoundingClientRect())
  );

  return (
    <div
      className={classnames.element("root")}
      style={{
        ...displayProperties.style,
        width: POPPER_WIDTH,
        height: POPPER_HEIGHT
      }}
    >
      {displayArrow && (
        <div
          style={displayProperties.arrowStyle}
          className={classnames.element("arrow", {
            left: displayProperties.arrowOrientation === ARROW_ORIENTATION.LEFT,
            right:
              displayProperties.arrowOrientation === ARROW_ORIENTATION.RIGHT
          })}
        />
      )}
      <div className={classnames.element("content")}>{props.children}</div>
    </div>
  );
};
Popper.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}Popper`);
Popper.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    content: PropTypes.string,
    arrow: PropTypes.string,
    arrowLeft: PropTypes.string,
    arrowRight: PropTypes.string
  }),
  anchorEl: PropTypes.object.isRequired,
  displayArrow: PropTypes.bool
};
Popper.defaultProps = {
  classes: {},
  displayArrow: true
};

export default withStyles(styles)(Popper);
