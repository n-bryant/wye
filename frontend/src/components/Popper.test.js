import { getPopperProps, constants, Popper } from "./Popper";

describe("Popper", () => {
  const props = {
    classes: {
      root: "root",
      content: "content",
      arrow: "arrow",
      arrowLeft: "arrowLeft",
      arrowRight: "arrowRight"
    },
    anchorEl: {
      getBoundingClientRect: jest.fn(() => ({
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        height: 0,
        width: 0
      }))
    },
    displayArrow: true,
    children: "foo"
  };

  let wrapper;
  beforeEach(() => {
    const displayProperties = {
      style: {},
      arrowStyle: {},
      arrowOrientation: constants.ARROW_ORIENTATION.LEFT
    };
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [displayProperties]);
    wrapper = shallow(<Popper {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set the arrow styles based on the arrow orientation value", () => {
    expect(
      wrapper.findWhere(n => n.hasClass(props.classes.arrowLeft)).length
    ).toBe(1);

    const displayProperties = {
      style: {},
      arrowStyle: {},
      arrowOrientation: constants.ARROW_ORIENTATION.RIGHT
    };
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [displayProperties]);
    const wrapperWithRightArrow = shallow(<Popper {...props} />);
    expect(
      wrapperWithRightArrow.findWhere(n => n.hasClass(props.classes.arrowRight))
        .length
    ).toBe(1);
  });
});

describe("getPopperProps", () => {
  let anchorRect = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    width: 250,
    height: 250
  };

  it("should return a style object with display none if there is no room on top", () => {
    window.innerHeight = constants.POPPER_HEIGHT - 1;
    expect(getPopperProps(anchorRect)).toMatchObject({
      style: {
        display: "none"
      }
    });
  });

  it("should set a left oriented arrow and appropriate styles if there is room to the right", () => {
    window.innerHeight = constants.POPPER_HEIGHT * 2;
    expect(getPopperProps(anchorRect)).toMatchObject({
      arrowOrientation: constants.ARROW_ORIENTATION.LEFT,
      arrowStyle: {
        top: `calc(${anchorRect.height}px / 4)`
      },
      style: {
        left: `${anchorRect.width + constants.POPPER_MARGIN}px`
      }
    });
  });

  it("should set a right oriented arrow and appropriate styles if there is no room to the right, but is room to the left", () => {
    window.innerHeight = constants.POPPER_HEIGHT * 2;
    window.innerWidth = constants.POPPER_WIDTH + 2;
    anchorRect = {
      ...anchorRect,
      left: constants.POPPER_WIDTH + constants.POPPER_MARGIN + 2,
      right: constants.POPPER_WIDTH + constants.POPPER_MARGIN + 1
    };
    expect(getPopperProps(anchorRect)).toMatchObject({
      arrowOrientation: constants.ARROW_ORIENTATION.RIGHT,
      arrowStyle: {
        top: `calc(${anchorRect.height}px / 4)`
      },
      style: {
        left: `-${constants.POPPER_WIDTH + constants.POPPER_MARGIN}px`
      }
    });
  });

  it("should set offsets and adjust styles if there is not room on the bottom", () => {
    window.innerHeight = constants.POPPER_HEIGHT + constants.HEADER_HEIGHT + 5;
    window.innerWidth = constants.POPPER_WIDTH + 2;
    anchorRect = {
      ...anchorRect,
      top: constants.POPPER_HEIGHT,
      left: constants.POPPER_WIDTH + constants.POPPER_MARGIN + 2,
      right: constants.POPPER_WIDTH + constants.POPPER_MARGIN + 1
    };
    expect(getPopperProps(anchorRect)).toMatchObject({
      arrowOrientation: "right",
      arrowStyle: { top: "calc(250px / 4 + -311px)" },
      style: { left: "-312px", top: 311 }
    });
  });
});
