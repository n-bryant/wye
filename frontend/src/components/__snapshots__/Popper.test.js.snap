// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper should render successfully 1`] = `
<div
  className="wye-Popper__root root"
  style={
    Object {
      "height": 354,
      "width": 300,
    }
  }
>
  <div
    className="wye-Popper__arrow wye-Popper__arrow--left arrow arrowLeft"
    style={Object {}}
  />
  <div
    className="wye-Popper__content content"
  >
    foo
  </div>
</div>
`;
